### PR TITLE
Hotfix/update on create

### PR DESCRIPTION
**Pull request by RWOverdijk**
This PR makes it so that:

* You can view the diff on a model
* You don't get afterUpdate to trigger on create
* You don't get an update at all, on create (because there's no diff)
* .save() only updates changed fields, in stead of everything
* There's no association cache anymore
* The original values are retrievable on the model.

### Notes to self
* ~~`.compareModelValues()` in `save.js` can probably go.~~
* ~~`.findPrimaryKey()` can probably go. Context has a primaryKey property.~~
* ~~Don't forget to set status back to clean after various save operations.~~

### CODE REVIEW

File: lib/waterline/model/lib/defaultMethods/save.js [hunk 1]
```diff
@@ -170,15 +146,26 @@ module.exports = function(context, proto, cb) {
...
       return cb(error);
     }
 
-    var obj = results.updateRecord[0].toObject();
+    if (dirtyProperties) {
+      proto.clean();
+    }
+
+    if (typeof results.updateRecord === 'undefined') {
+      results.updateRecord = proto;
```
Comment (RWOverdijk) [inline lib/waterline/model/lib/defaultMethods/save.js]: Not sure about this. What should be done here?

**Comment by RWOverdijk** [comment]
Tests currently fail. The first two errors are because of [this piece of code](https://github.com/balderdashy/waterline/blob/master/test/unit/model/association.add.hasMany.id.js#L31). The model didn't change. This PR is now responsible for stopping .save() from updating the model if nothing changed. So, .updated is empty... So you get this error. This means that the reason these tests fail, is because they relied on .update() being called even if there weren't any changes.

TL;DR; I will change the tests to make sure they reflect the truth.

As to the rest of the tests, I'll fix those after.

**Comment by RWOverdijk** [comment]
Now that the build is succeeding and all tests for the changes have been written... Let's talk. It's a pretty massive change.

**Comment by RWOverdijk** [comment]
Now that the build is succeeding and all tests for the changes have been written... Let's talk. It's a pretty massive change.

**Comment by dmarcelino** [comment]
Nice... I'll need some time to digest this, meanwhile lets cc @devinivy, @particlebanana and @tjwebb.

**Comment by RWOverdijk** [comment]
Alrighty. I can confirm this solves my initial issue by the way. afterUpdate isn't being called on create anymore.

**Comment by dmarcelino** [comment]
So, for reference, this replaces PR #710, and fixes #1035 and #683, right?

**Comment by RWOverdijk** [comment]
@dmarcelino Correct. As well as reduce the number of queries made most of the time and removing unneeded code (mutatedModels, associationCache). But it also no longer calls afterUpdate on create so, technically it's a BC break.

**Comment by dmarcelino** [comment]
So many good things, it must be an early xmas! :christmas_tree:

> But it also no longer calls afterUpdate on create so, technically it's a BC break.

I'm not sure it qualifies as breaking change since afterUpdate on create is a bug and not an intended (or even documented) feature. In other words, I doubt people would code expecting that behaviour, but we should discuss it further.

**Comment by RWOverdijk** [comment]
I've also added .getPrimaryKey, but that can be taken out.

Pretty sure that reserving .clean and .dirty on models should be considered a BC break though haha. People might be using those.

**Comment by dmarcelino** [comment]
You're right, I'm pretty sure they are. We should should think of alternatives that will not colide with user's models. Something like `isCleanRecord()` and `isDirtyRecord()` should minimize collisions...

**Comment by RWOverdijk** [comment]
@dmarcelino Probably .getDirtyProperties() and .setClean() then. isCleanRecord() currently equals to `.dirty() === false`. 

I personally like . unsynced() or .getUnsynced() in stead of .dirty(), and .synced() or .setSynced() for .clean(). But I took the naming from the other PR.

**Comment by particlebanana** [comment]
@RWOverdijk awesome work! This is great. We definitely need to change the `.dirty` and `.clean` method names though. 

Initially I was thinking we could add underscores to represent private methods but having access to the dirty properties could be pretty useful. I'm not sure what the use case for having `.clean()` be public is though. 

Rails uses the `changed` wording so we could do something along those lines - `.getChanged()` would return the attributes that have changed then make `._resetSynced()` a private method. Not as friendly as `.dirty()` and `.clean()` but maybe less confusing? Just a thought.

**Comment by RWOverdijk** [comment]
@particlebanana .getChanged() makes it implicit and simple, with fewer chance of naming collision. I like it. Perhaps use the same wording for .clean()? .applyChanges()? Or .submitState().. Something. It's not really a reset, because you're saying that the current state has been synced as is now the currentState. Something like .acceptCurrentState(properties)?

**Comment by RWOverdijk** [comment]
cool cool. Needs rebasing now

**Comment by dmarcelino** [comment]
> cool cool. Needs rebasing now

:disappointed:  We are closing `0.10` and preparing one last hotfix. Afterwards our focus will be in `0.11`

For reference, this should also fix #1015.

**Comment by RWOverdijk** [comment]
@dmarcelino Just let me know when to rebase this stuff.

**Comment by dmarcelino** [comment]
Hey @RWOverdijk, I believe your PR was broken by all the linting that took place, there are still a few linting rules breaking, https://github.com/balderdashy/waterline/pull/1084#issuecomment-121133140:

> screenshot from 2015-07-14 00 56 32

So there is some risk that further linting will again break your PR :confused: 

BTW, we forked `0.10` from master for hotfixes and master is now dedicated to `0.11`.

**Comment by RWOverdijk** [comment]
@dmarcelino I'll sit patiently.

**Comment by RWOverdijk** [comment]
@dmarcelino Come on man.

**Comment by weyert** [comment]
Any news whether this PR gets merged into master any time soon?

**Comment by devinivy** [comment]
I think it's safe to rebase this before linting continues.

**Comment by RWOverdijk** [comment]
Does it even make sense at this point? Is it safe?

**Comment by RWOverdijk** [comment]
Fuck it.

**Comment by devinivy** [comment]
@RWOverdijk :fireworks:

**Comment by saschagehlich** [comment]
@RWOverdijk Why did this never get merged?

**Comment by RWOverdijk** [comment]
@saschagehlich Because I got tired of waiting and having to keep rebasing. I went full bender and built my [own ORM](https://wetland.spoonx.org). With blackjack, and hoo... Ladies.

**Comment by RWOverdijk** [comment]
@saschagehlich There's a [sails hook](https://github.com/SpoonX/sails-hook-wetland) but I haven't tested it with 1.0 yet. I'm using it in production. :)

It doesn't work with document based storage. It's an ORM, not ODM. :)

**Comment by saschagehlich** [comment]
NP, been having problems with MongoDB and CouchDB in larger setups anyways. Postgres ftw!

**Comment by RWOverdijk** [comment]
Haha I agree. 👍  Let's not go too offtopic here. If you have questions you can hit me up on gitter. https://gitter.im/spoonx/dev
